### DISCUSSION TITLE
Form with dynamic objects

### DISCUSSION THREAD

**Discussion by arstiefel** [react-hook-form/react-hook-form#8120] [Q&A]
I am creating a very large form using 7.29.0. In my form one of my objects can have 1 to many children. One of my dropdowns can dynamically create a new entry in the object. It's similar to field array but it is an object not field array. When trying to use setValue to create the new object I have tried both looping through each field as well as setting the entire value. It randomly creates an undefined object along with the correct new object.

Additionally, when I use shouldDirty to be true it is incredibly slow to react.

It looks like my useWatch is getting triggered a lot of times. I am watching the object that contains the changing object. Is there a way to do a deep comparison of the objects?
The documentation says `useWatch result is optimised for render phase instead of useEffect's deps, to detect value update you may want to use an external custom hook for value comparison.` but I'm not sure how to do that.

Basically, I feed the result of useWatch into useMemo. I am using a deep compare in that useMemo but the useWatch is still causing a lot of rerenders anyway

**Comment by bluebill1049** [comment]
can you share a codesandbox? and show care how are you using the `useWatch` hook

**Reply by arstiefel** [reply]
I am not able to use sandbox, unfortunately. The code snippet is basically 

```
const products = useWatch({
    name: `a.b.c.products`
  });
const activeProducts = useMemo(
  () => getActiveProducts(products),
  [products]
);

....

return (
  <Select
        value={activeProducts[0].productId}
        onChange={onChangeProduct}
        options={issueTypeOptions}
  />
)
```

where products is an object `{[key]: {...}}`

I did try wrapping the useWatch in a deep compare function I still see a big lag

```
export function useDeepCompareMemoize<T>(value: T) {
  const ref = useRef<T>(value);
  const signalRef = useRef<number>(0);

  if (!deepEqual(value, ref.current)) {
    ref.current = _.cloneDeep(value);
    signalRef.current += 1;
  }

  return useMemo(() => ref.current, [signalRef.current]);
}
```
